src/main/java/com/pnudev/librarysystem/service/BookService.java

```diff
@@ -0,0 +1,141 @@
+import java.nio.file.StandardCopyOption;
+import java.util.List;
+import java.util.UUID;
+
+
+@RequiredArgsConstructor
+@Service
+public class BookService {
+
+    @Value("${file.upload-dir}/books")
+    private String uploadDir;
+
+    @Value("${file.image-types}")
+    private List<String> imageTypes;
+
+    private final BookRepository bookRepository;
+    private final BookMapper bookMapper;
+
+
+    public Long createBook(RequestBookDTO requestBookDTO)  {
+        Book book = bookMapper.toEntity(requestBookDTO);
+        Book savedBook = bookRepository.save(book);
+        return savedBook.getId();
+    }
+
+    public BookDTO updateBook(Long id, RequestBookDTO requestBookDTO) {
+        Book book = bookRepository.findById(id)
+                .orElseThrow(() -> new EntityNotFoundException("Book with id: %d not found".formatted(id)));
+
+        bookMapper.updateBookFromRequestDTO(requestBookDTO, book);
+
+        return bookMapper.toDTO(bookRepository.save(book));
+    }
+
+    public void deleteBookById(Long id) {
+        Book book = bookRepository.findById(id)
+                .orElseThrow(() -> new EntityNotFoundException("Book with id: %d not found".formatted(id)));
+
+        if(!book.getBorrowings().isEmpty()) {
+            throw new DeleteFailedException("Book with id: %d is used".formatted(id));
+        }
+
+        bookRepository.deleteById(id);
+    }
+
+    public Page<BookDTO> searchBook(String title, String authorLastName, String categoryName, Pageable pageable) {
+        Specification<Book> specification = Specification.where(null);
+
+        if (title != null && !title.isEmpty()) {
+            specification = specification.and(SpecificationUtils.<Book>fieldContainsIgnoreCase("title", title));
+        }
+        if (authorLastName != null && !authorLastName.isEmpty()){
+            specification = specification.and(
+                    SpecificationUtils.<Book>joinTableFieldContainsIgnoreCase("authors", "lastName", authorLastName)
+            );
+        }
+        if (categoryName != null && !categoryName.isEmpty()){
+            specification = specification.and(
+                    SpecificationUtils.<Book>joinTableFieldContainsIgnoreCase("categories", "name", categoryName)
+            );
+        }
+
+        Page<Book> page = bookRepository.findAll(specification, pageable);
+        return page.map(bookMapper::toDTO);
+    }
+
+    public BookDTO getBookById(Long id) {
+        Book book = bookRepository.findById(id)
+                .orElseThrow(() -> new EntityNotFoundException("Book with id: %d not found".formatted(id)));
+        return bookMapper.toDTO(book);
+    }
+
+    public String uploadCoverImage(MultipartFile file) {
+        if(file.isEmpty()) {
+            throw new EmptyFileException("File cannot be empty");
+        }
+
+        String originalFileName = file.getOriginalFilename();
+        String fileExtension = originalFileName.substring(originalFileName.lastIndexOf(".")+1);
```

Comment:
StringUtils.substringAfterLast for the file extension

Reply:
Done